### PR TITLE
Cassandra 17333 trunk

### PR DESCRIPTION
**Pull request by bereng**

### CODE REVIEW

**Comment by bereng** [comment]
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/587/workflows/b07dbd6a-725e-433f-80dc-b32f5f3cbcf3)
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/587/workflows/efd480e6-b8ad-4ac1-8dcb-7fcd841814f1)

File: src/java/org/apache/cassandra/auth/IAuthorizer.java [hunk 1]
```diff
@@ -62,12 +62,13 @@ default boolean requireAuthorization()
      * @param permissions Set of permissions to grant.
      * @param resource Resource on which to grant the permissions.
      * @param grantee Role to which the permissions are to be granted.
+     * @return the permissions that have been successfully granted.
```
Comment (adelapena) [inline src/java/org/apache/cassandra/auth/IAuthorizer.java]: We might mention that this is different to the `permissions` attribute, for example:
```suggestion
     * @return the permissions that have been successfully granted, comprised by the requested permissions excluding
     * those permissions that were already granted.
```

File: src/java/org/apache/cassandra/auth/IAuthorizer.java [hunk 2]
```diff
@@ -80,12 +81,13 @@ void grant(AuthenticatedUser performer, Set<Permission> permissions, IResource r
      * @param permissions Set of permissions to revoke.
      * @param revokee Role from which to the permissions are to be revoked.
      * @param resource Resource on which to revoke the permissions.
+     * @return the permissions that have been successfully revoked.
```
Comment (adelapena) [inline src/java/org/apache/cassandra/auth/IAuthorizer.java]: As before, we could mention that this is different to the `permissions` attribute, for example:
```suggestion
     * @return the permissions that have been successfully revoked, comprised by the requested permissions excluding
     * those permissions that were already not granted.
```

File: src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java [hunk 3]
```diff
@@ -39,7 +43,25 @@ public GrantPermissionsStatement(Set<Permission> permissions, IResource resource
...
+
+        // We want to warn the client if all the specified permissions have not been granted and the client did
+        // not specified ALL in the query.
+        if (!granted.equals(permissions) && !permissions.equals(Permission.ALL))
+        {
+            // We use a TreeSet to guarantee the order for testing
+            String permissionsStr = new TreeSet<>(permissions).stream()
+                                                              .filter(permission -> !granted.contains(permission))
+                                                              .map(Permission::name)
+                                                              .collect(Collectors.joining(", "));
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java]: Given that we are creating a stream, we could save us the `TreeSet` by just ordering the stream, for example:
```suggestion
            String permissionsStr = permissions.stream()
                                               .filter(permission -> !granted.contains(permission))
                                               .sorted(Permission::compareTo) // guarantee the order for testing
                                               .map(Permission::name)
                                               .collect(Collectors.joining(", "));
```

File: src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java [hunk 4]
```diff
@@ -41,7 +45,25 @@ public RevokePermissionsStatement(Set<Permission> permissions, IResource resourc
...
+
+        // We want to warn the client if all the specified permissions have not been revoked and the client did
+        // not specified ALL in the query.
+        if (!revoked.equals(permissions) && !permissions.equals(Permission.ALL))
+        {
+            // We use a TreeSet to guarantee the order for testing
+            String permissionsStr = new TreeSet<>(permissions).stream()
+                                                              .filter(permission -> !revoked.contains(permission))
+                                                              .map(Permission::name)
+                                                              .collect(Collectors.joining(", "));
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java]: As before, we can use `Stream#sort` instead of the `TreeSet`:
```suggestion
            String permissionsStr = permissions.stream()
                                               .filter(permission -> !revoked.contains(permission))
                                               .sorted(Permission::compareTo) // guarantee the order for testing
                                               .map(Permission::name)
                                               .collect(Collectors.joining(", "));
```

File: src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java [hunk 5]
```diff
@@ -39,7 +43,25 @@ public GrantPermissionsStatement(Set<Permission> permissions, IResource resource
 
     public ResultMessage execute(ClientState state) throws RequestValidationException, RequestExecutionException
     {
-        DatabaseDescriptor.getAuthorizer().grant(state.getUser(), permissions, resource, grantee);
+        IAuthorizer authorizer = DatabaseDescriptor.getAuthorizer();
+        Set<Permission> granted = authorizer.grant(state.getUser(), permissions, resource, grantee);
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java:46]: Has this been split to break the long line?

File: src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java [hunk 6]
```diff
@@ -41,7 +45,25 @@ public RevokePermissionsStatement(Set<Permission> permissions, IResource resourc
 
     public ResultMessage execute(ClientState state) throws RequestValidationException, RequestExecutionException
     {
-        DatabaseDescriptor.getAuthorizer().revoke(state.getUser(), permissions, resource, grantee);
+        IAuthorizer authorizer = DatabaseDescriptor.getAuthorizer();
+        Set<Permission> revoked = authorizer.revoke(state.getUser(), permissions, resource, grantee);
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java:48]: Has this been split to break the long line?

File: src/java/org/apache/cassandra/auth/CassandraAuthorizer.java [hunk 7]
```diff
@@ -92,18 +96,41 @@ public CassandraAuthorizer()
...
+
+        nonExistingPermissions.removeAll(getExistingPermissions(roleName, resourceName, permissions));
+        if (!nonExistingPermissions.isEmpty())
+        {
+            modifyRolePermissions(nonExistingPermissions, resource, grantee, "+");
+            addLookupEntry(resource, grantee);
+            grantedPermissions.addAll(nonExistingPermissions);
+        }
+
+        return grantedPermissions;
```
Comment (adelapena) [inline src/java/org/apache/cassandra/auth/CassandraAuthorizer.java]: I think `grantedPermissions` is redundant, since it's an exact copy of `nonExistingPermissions`. We can probably simplify to something like:
```suggestion
        Set<Permission> existingPermissions = getExistingPermissions(roleName, resourceName, permissions);
        Set<Permission> nonExistingPermissions = Sets.difference(permissions, existingPermissions);

        if (!nonExistingPermissions.isEmpty())
        {
            modifyRolePermissions(nonExistingPermissions, resource, grantee, "+");
            addLookupEntry(resource, grantee);
        }

        return nonExistingPermissions;
```

File: src/java/org/apache/cassandra/auth/CassandraAuthorizer.java [hunk 8]
```diff
@@ -92,18 +96,41 @@ public CassandraAuthorizer()
...
+        Set<Permission> existingPermissions = getExistingPermissions(roleName, resourceName, permissions);
+
+        if (!existingPermissions.isEmpty())
+        {
+            modifyRolePermissions(existingPermissions, resource, revokee, "-");
+            removeLookupEntry(resource, revokee);
+            revokedPermissions.addAll(existingPermissions);
+        }
+
+        return revokedPermissions;
```
Comment (adelapena) [inline src/java/org/apache/cassandra/auth/CassandraAuthorizer.java]: As before, I think that either `revokedPermissions` or `existingPermissions` is redundant, since one is a copy of the other. We can simplify to:
```suggestion
        Set<Permission> existingPermissions = getExistingPermissions(roleName, resourceName, permissions);

        if (!existingPermissions.isEmpty())
        {
            modifyRolePermissions(existingPermissions, resource, revokee, "-");
            removeLookupEntry(resource, revokee);
        }

        return existingPermissions;
```

File: test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java [hunk 9]
```diff
@@ -357,4 +360,29 @@ public void testGrantedAllTables() throws Throwable
...
+
+        executeNet("CREATE KEYSPACE revoke_yeah WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}");
+        executeNet("CREATE TABLE revoke_yeah.t1 (id int PRIMARY KEY, val text)");
+        executeNet("CREATE USER '" + user + "' WITH PASSWORD '" + pass + "'");
+
+        ResultSet res = executeNet("REVOKE CREATE ON KEYSPACE revoke_yeah FROM " + user);
+        assertWarningsContain(res.getExecutionInfo().getWarnings(), "Role '" + user + "' was not granted CREATE on <keyspace revoke_yeah>");
+
+        res = executeNet("GRANT SELECT ON KEYSPACE revoke_yeah TO " + user);
+        assertTrue(res.getExecutionInfo().getWarnings().size() == 0);
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java]: Nit:
```suggestion
        assertTrue(res.getExecutionInfo().getWarnings().isEmpty());
```

File: test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java [hunk 10]
```diff
@@ -357,4 +360,29 @@ public void testGrantedAllTables() throws Throwable
...
+    public void testWarnings() throws Throwable
+    {
+        useSuperUser();
+
+        executeNet("CREATE KEYSPACE revoke_yeah WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}");
+        executeNet("CREATE TABLE revoke_yeah.t1 (id int PRIMARY KEY, val text)");
+        executeNet("CREATE USER '" + user + "' WITH PASSWORD '" + pass + "'");
+
+        ResultSet res = executeNet("REVOKE CREATE ON KEYSPACE revoke_yeah FROM " + user);
+        assertWarningsContain(res.getExecutionInfo().getWarnings(), "Role '" + user + "' was not granted CREATE on <keyspace revoke_yeah>");
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java:374]: Probably it doesn't matter for this case, but I wonder if we should check that the emitted warnings are exactly the expected ones, so check that there aren't any other unexpected warnings. Perhaps we could have (quite trivial) utility methods for that on `CQLTester`, for example:
```java
protected static void assertWarningsEquals(ResultSet rs, String... messages)
{
    assertWarningsEquals(rs.getExecutionInfo().getWarnings(), messages);
}

protected static void assertWarningsEquals(List<String> warnings, String... messages)
{
    Assert.assertNotNull(warnings);
    Assertions.assertThat(messages).hasSameElementsAs(warnings);
}
```

File: build.xml [hunk 11]
```diff
@@ -2131,26 +2148,16 @@
...
- 	 <fileset dir="build/lib/jars">
-  	    <include name="**/*.jar" />
-  	 </fileset>
-     <fileset dir="build/test/lib/jars">
-        <include name="**/*.jar" />
-     </fileset>
-  	</path>
-  	<property name="eclipse-project-libs" refid="eclipse-project-libs-path"/>
-      <taskdef name="echoeclipseprojectslibs" classname="org.apache.cassandra.anttasks.EchoEclipseProjectLibs" classpath="${test.classes}">
+    <taskdef name="echoeclipseprojectslibs" classname="org.apache.cassandra.anttasks.EchoEclipseProjectLibs" classpath="${test.classes}">
```
Comment (adelapena) [inline build.xml]: What's the purpose of these changes?

Comment (bereng) [reply build.xml]: At the time eclipse was broken and I needed this to fix it. You can see it is in an isolated commit. I will rebase.

File: src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java (hunk 5 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java:46]: For no specific reason. Reads better imo and prevents a long line. You'd like it the old way better?

File: test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java (hunk 10 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java:374]: That is a good idea at face value. The problem is that warnings are accumulative in nature. Sometime sbdy is going to add a new warning somewhere and then tests will start to fail and you'll have to fix them. So I am +1 in adding these methods for the occasional test where an exact match can be guaranteed but not here. Wdyt?

**Comment by bereng** [comment]
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/601/workflows/bd427f19-04c4-4c3c-bbd4-541eb80bd4d1)
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/601/workflows/595ecd3d-1bf6-4cff-8a49-6dc8d304b313)

**Comment by bereng** [comment]
@adelapena I think all your comments have been addressed.

File: src/java/org/apache/cassandra/auth/CassandraAuthorizer.java [hunk 12]
```diff
@@ -28,12 +28,16 @@
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Table;
+import com.google.common.collect.Sets;
+
+import org.apache.commons.lang3.ArrayUtils;
```
Comment (adelapena) [inline src/java/org/apache/cassandra/auth/CassandraAuthorizer.java]: Nit: unused import

File: src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java [hunk 13]
```diff
@@ -18,16 +18,20 @@
 package org.apache.cassandra.cql3.statements;
 
 import java.util.Set;
+import java.util.TreeSet;
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java]: Nit: unused import

File: src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java [hunk 14]
```diff
@@ -39,7 +43,25 @@ public GrantPermissionsStatement(Set<Permission> permissions, IResource resource
 
     public ResultMessage execute(ClientState state) throws RequestValidationException, RequestExecutionException
     {
-        DatabaseDescriptor.getAuthorizer().grant(state.getUser(), permissions, resource, grantee);
+        IAuthorizer authorizer = DatabaseDescriptor.getAuthorizer();
+        Set<Permission> granted = authorizer.grant(state.getUser(), permissions, resource, grantee);
+
+        // We want to warn the client if all the specified permissions have not been granted and the client did
+        // not specified ALL in the query.
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java]:
```suggestion
        // not specify ALL in the query.
```

File: src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java [hunk 15]
```diff
@@ -18,16 +18,20 @@
 package org.apache.cassandra.cql3.statements;
 
 import java.util.Set;
+import java.util.TreeSet;
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java]: Nit: unused import

File: src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java [hunk 16]
```diff
@@ -41,7 +45,25 @@ public RevokePermissionsStatement(Set<Permission> permissions, IResource resourc
 
     public ResultMessage execute(ClientState state) throws RequestValidationException, RequestExecutionException
     {
-        DatabaseDescriptor.getAuthorizer().revoke(state.getUser(), permissions, resource, grantee);
+        IAuthorizer authorizer = DatabaseDescriptor.getAuthorizer();
+        Set<Permission> revoked = authorizer.revoke(state.getUser(), permissions, resource, grantee);
+
+        // We want to warn the client if all the specified permissions have not been revoked and the client did
+        // not specified ALL in the query.
```
Comment (adelapena) [inline src/java/org/apache/cassandra/cql3/statements/RevokePermissionsStatement.java]:
```suggestion
        // not specify ALL in the query.
```

File: test/unit/org/apache/cassandra/auth/StubAuthorizer.java [hunk 17]
```diff
@@ -42,34 +42,55 @@ public void clear()
...
+                                 Set<Permission> permissions,
+                                 IResource resource,
+                                 RoleResource grantee) throws RequestValidationException, RequestExecutionException
     {
         Pair<String, IResource> key = Pair.create(grantee.getRoleName(), resource);
-        Set<Permission> perms = userPermissions.get(key);
-        if (null == perms)
+        Set<Permission> oldPermissions = userPermissions.get(key);
+
+        if (oldPermissions == null)
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/auth/StubAuthorizer.java]: Not a big deal for a test class, but we could simplify this method a bit by using `Set#computeIfAbsent` and Guava's sets:
```java
{
    Pair<String, IResource> key = Pair.create(grantee.getRoleName(), resource);
    Set<Permission> oldPermissions = userPermissions.computeIfAbsent(key, k -> Collections.emptySet());
    Set<Permission> nonExisting = Sets.difference(permissions, oldPermissions);

    if (!nonExisting.isEmpty())
        userPermissions.put(key, Sets.union(oldPermissions, nonExisting));

    return nonExisting;
}
```

File: test/unit/org/apache/cassandra/auth/StubAuthorizer.java [hunk 18]
```diff
@@ -42,34 +42,55 @@ public void clear()
...
-                       RoleResource revokee) throws RequestValidationException, RequestExecutionException
+    public Set<Permission> revoke(AuthenticatedUser performer,
+                                  Set<Permission> permissions,
+                                  IResource resource,
+                                  RoleResource revokee) throws RequestValidationException, RequestExecutionException
     {
         Pair<String, IResource> key = Pair.create(revokee.getRoleName(), resource);
-        Set<Permission> perms = userPermissions.get(key);
-        if (null != perms)
+        Set<Permission> oldPermissions = userPermissions.get(key);
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/auth/StubAuthorizer.java]: As before, perhaps we can simplify the method to:
```java
{
    Pair<String, IResource> key = Pair.create(revokee.getRoleName(), resource);
    Set<Permission> oldPermissions = userPermissions.computeIfAbsent(key, k -> Collections.emptySet());
    Set<Permission> existing = Sets.intersection(permissions, oldPermissions);

    if (existing.isEmpty())
        userPermissions.remove(key);
    else
        userPermissions.put(key, Sets.difference(oldPermissions, existing));

    return existing;
}
```

File: src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java (hunk 5 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java:46]: Fine with me.

File: test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java (hunk 10 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/auth/GrantAndRevokeTest.java:374]: I don't see what other warnings could be issued by those specific queries, but I guess it might happen. I was thinking on some issues with duplicated warnings that we have had in the past. But I'm fine with leaving it as it is.

**Review by adelapena — COMMENTED** [review]
Looks good to me, I have just left a few additional suggestions about nits. Could we start a repeated run of the new test to verify that it's stable, just in case?

File: test/unit/org/apache/cassandra/auth/StubAuthorizer.java (hunk 17 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/auth/StubAuthorizer.java]: Nice I like it.

**Comment by bereng** [comment]
Let's see how it goes...
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/602/workflows/f95a1371-8650-4b4f-956a-780a408604e2)
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/602/workflows/cca7640d-1d24-4164-9963-9a7b20e1e7ef)

**Comment by adelapena** [comment]
It seems that some of those runs hit the ccm problem that we have had lately. Here are runs using the patched ccm version:
[j8](https://app.circleci.com/pipelines/github/adelapena/cassandra/1315/workflows/b0b424e4-ee36-4e2a-ad6e-4b37310dcbfc)
[j11](https://app.circleci.com/pipelines/github/adelapena/cassandra/1315/workflows/6e270574-82ce-47c0-8c03-1efd350e3c83)
All looks good, included the 400 iterations of the new test.

**Review by adelapena — APPROVED** [review]
